### DISCUSSION TITLE
How does escaping in i3-msg exec work?

### DISCUSSION THREAD

**Discussion by meator** [i3/i3#6051] [Q&A]
Hey. I'm trying to comprehend i3's escaping rules for the `exec` command of i3 IPC. I am integrating i3 IPC into a program I maintain and I need to implement perfect forwarding of the command line.

This is how I understand the unquoting process when `exec`'s argument is withing `""` quotes (like this: `i3-msg 'exec "true"'`):
1. replace occurrences of `\\"` with `"` [as noted in documentation](https://i3wm.org/docs/userguide.html#exec_quoting)
2. remove the beginning and end `"`
3. pass the resulting string to `sh -c <CMDLINE>` as a single argument (I haven't verified whether `sh` itself is executed or whether some other shell is run, but that shouldn't matter much)

I have found no mention of escaping `\` itself. This means that either the documentation is lacking or that it isn't implemented. If it isn't implemented, it could be considered a flaw in the escaping system.

---

Let's say that I want to make `i3` run the following program:

```
prog \\
```

i.e. run program with argv0 `prog` (or `\x70\x72\x6F\x67` in hex to be precise) and with argv1 `\\` (or `\x5C\x5C` in hex).

Because `i3` will pass it through a shell, the backslashes have to be doubled. I would expect the following command:

```
i3-msg 'exec prog \\\\'
```

to provide the following input to `prog`'s argv1:

```
\\
```

it provides

```
\
```

instead. The following command:

```
i3-msg 'exec prog \\\\\\\\'
```

indeed does provide `prog` with `\\` as argv0. Note that

```
i3-msg 'exec prog \\\\\\\'
```

also works, which is even weirder. I don't know why it works. When considering [the exec quoting section of the user manual](https://i3wm.org/docs/userguide.html#exec_quoting), which is to my knowledge the only place escaping is documented at all, the previous outputs do not make sense.

When I simulate the shell unescaping the arguments using `echo`:

```
echo -E prog \\\\\\\\
```

`prog` receives

```
\\\\
```

which is double the backslashes the program receives when run via `i3-msg`.

I have examined `i3-msg` with strace to inspect the writes to the unix sockets. I have also manually interacted with it to make sense of this.

I have many more questions about the escaping rules of i3 IPC, but this comment is already long and I believe that I will be able to answer the rest of my questions if I figure out the basics I am asking about in this discussion.

**Comment by meator** [comment]
I did some further testing. It looks like the double backslash rule for double quotes is just false. I have caused some of my confusion myself, but some parts are still unclear. It is important to keep in mind that there are two layers of unescaping:

1. done by i3 itself
2. done by the shell that i3 executes

Some of my confusion was caused by forgetting the shell. But I believe that the only place the escaping behavior is documented is just false. I have figured things out by trial and error.

The i3 IPC documentation doesn't document escaping rules. I assume that escaping rules in [the userguide](https://i3wm.org/docs/userguide.html#exec_quoting) should be followed, but maybe they are true in the i3 config file only.

**Reply by orestisfl** [reply]
i3 escapes either `\\` or `\"`, otherwise it treats `\<anything>` as normal string chars.

Here's a list of i3 `exec` commands followed by what actually gets executed:
```
IPC: received: *exec notify-send \*
executing: notify-send \
IPC: received: *exec notify-send \\*
executing: notify-send \
IPC: received: *exec notify-send \\\*
executing: notify-send \\
IPC: received: *exec notify-send \\\\*
executing: notify-send \\
IPC: received: *exec notify-send \\\\\*
executing: notify-send \\\
IPC: received: *exec notify-send \\\\\\*
executing: notify-send \\\
IPC: received: *exec notify-send \\\\\\\*
executing: notify-send \\\\
IPC: received: *exec notify-send \\\\\\\\*
executing: notify-send \\\\

IPC: received: *exec notify-send "*
executing: notify-send "
IPC: received: *exec notify-send \"*
executing: notify-send "
IPC: received: *exec notify-send \\"*
executing: notify-send \"
IPC: received: *exec notify-send \\\"*
executing: notify-send \"
```

I got that by running `i3-dump-log -f | rg -i 'ipc.*received|executing:.*notify.send'`. `*` is used as a delimiter for the logs.

As you mention in this comment, i3 passes everything after `executing: ` to `sh -c`, specifically `argv[0]="/bin/sh"`, `argv[1]="-c"`, `argv[2]=<the command>`.

**Reply by meator** [reply]
OK, thanks!

Why does the user guide make mention of `\\"`? Is there a third layer of escaping applied to the config file which resolves it to `\"`, which gets resolved by the IPC escaper to `"` which then gets passed to the shell (which will probably fail)? I assume that this mechanism, if it exists, isn't employed in raw IPC communication, so I am not affected by it.
